File: sdks/rust/test-optimization-sdk/src/test_optimization/test_optimization.rs
// Unless explicitly stated otherwise all files in this repository are licensed
// under the Apache License Version 2.0.
// This product includes software developed at Datadog (https://www.datadoghq.com/).
// Copyright 2025 Datadog, Inc.

//! Test optimization module for managing the test optimization library
//!
//! This module provides functionality for initializing and shutdown the library.
//! Also access to the backend features.

use crate::test_optimization::lib::*;
use crate::test_optimization::utils::*;
use std::collections::HashMap;
use std::ffi::{c_char, CStr, CString};
use std::ptr::null_mut;

#[derive(Debug, Clone)]
/// Represents the settings for a test session
pub struct Settings {
    /// Whether code coverage is enabled
    #[allow(dead_code)]
    pub code_coverage: bool,
    /// Settings for early flake detection
    #[allow(dead_code)]
    pub early_flake_detection: EfDSettings,
    /// Whether flaky test retries are enabled
    #[allow(dead_code)]
    pub flaky_test_retries_enabled: bool,
    /// Whether intelligent test runner is enabled
    #[allow(dead_code)]
    pub itr_enabled: bool,
    /// Whether git integration is required
    #[allow(dead_code)]
    pub require_git: bool,
    /// Whether test skipping is enabled
    #[allow(dead_code)]
    pub tests_skipping: bool,
    /// Whether known tests tracking is enabled
    #[allow(dead_code)]
    pub known_tests_enabled: bool,
    /// Settings for test management
    #[allow(dead_code)]
    pub test_management: TestManagementSettings,
}

#[derive(Debug, Clone)]
/// Settings for early flake detection
pub struct EfDSettings {
    /// Whether early flake detection is enabled
    #[allow(dead_code)]
    pub enabled: bool,
    /// Settings for slow test retries
    #[allow(dead_code)]
    pub slow_test_retries: EfdSlowTestRetriesSettings,
    /// Threshold for faulty session detection
    #[allow(dead_code)]
    pub faulty_session_threshold: i32,
}

#[derive(Debug, Clone)]
/// Settings for slow test retries in early flake detection
pub struct EfdSlowTestRetriesSettings {
    /// Number of retries for 5-minute tests
    #[allow(dead_code)]
    pub five_m: i32,
    /// Number of retries for 30-second tests
    #[allow(dead_code)]
    pub thirty_s: i32,
    /// Number of retries for 10-second tests
    #[allow(dead_code)]
    pub ten_s: i32,
    /// Number of retries for 5-second tests
    #[allow(dead_code)]
    pub five_s: i32,
}

#[derive(Debug, Clone)]
/// Settings for flaky test retries
pub struct FlakyTestRetriesSettings {
    /// Number of retries for flaky tests
    #[allow(dead_code)]
    pub retry_count: i32,
    /// Total number of retries allowed
    #[allow(dead_code)]
    pub total_retry_count: i32,
}

#[derive(Debug, Clone)]
/// Settings for test management
pub struct TestManagementSettings {
    /// Whether test management is enabled
    #[allow(dead_code)]
    pub enabled: bool,
    /// Number of retries for attempt-to-fix operations
    #[allow(dead_code)]
    pub attempt_to_fix_retries: i32,
}

#[derive(Debug, Clone)]
/// Represents a skippable test
pub struct SkippableTest {
    /// Name of the test suite
    #[allow(dead_code)]
    pub suite_name: String,
    /// Name of the test
    #[allow(dead_code)]
    pub test_name: String,
    /// Test parameters
    #[allow(dead_code)]
    pub parameters: String,
    /// Custom configurations in JSON format
    #[allow(dead_code)]
    pub custom_configurations_json: String,
}

#[derive(Debug, Clone)]
/// Represents a test managed by the test management system
pub struct TestManagementTest {
    /// Name of the module
    #[allow(dead_code)]
    pub module_name: String,
    /// Name of the test suite
    #[allow(dead_code)]
    pub suite_name: String,
    /// Name of the test
    #[allow(dead_code)]
    pub test_name: String,
    /// Whether the test is quarantined
    #[allow(dead_code)]
    pub quarantined: bool,
    /// Whether the test is disabled
    #[allow(dead_code)]
    pub disabled: bool,
    /// Whether the test is attempt-to-fix
    #[allow(dead_code)]
    pub attempt_to_fix: bool,
}

/// Language name for the test session
pub(in crate::test_optimization) static LANGUAGE_NAME: &str = "rust";
/// Runtime name for the test session
pub(in crate::test_optimization) static RUNTIME_NAME: &str = "rustc";

#[derive(Debug, Clone)]
/// Represents a test session
pub struct TestOptimization;

impl TestOptimization {
    /// Get the runtime version
    #[allow(dead_code)]
    pub fn runtime_version() -> String {
        rustc_version_runtime::version().to_string()
    }

    /// Initialize the test optimization library
    #[allow(dead_code)]
    pub fn init() -> bool {
        Self::init_with_values(
            LANGUAGE_NAME,
            RUNTIME_NAME,
            Self::runtime_version(),
            None::<&str>,
            false,
        )
    }

    /// Initialize the test optimization library with a working directory
    #[allow(dead_code)]
    pub fn init_with_working_dir(working_dir: &str) -> bool {
        Self::init_with_values(
            LANGUAGE_NAME,
            RUNTIME_NAME,
            Self::runtime_version(),
            Some(working_dir),
            false,
        )
    }

    /// Initialize the test optimization library with a mock tracer
    #[allow(dead_code)]
    pub fn init_mock() -> bool {
        Self::init_with_values(
            LANGUAGE_NAME,
            RUNTIME_NAME,
            Self::runtime_version(),
            None::<&str>,
            true,
        )
    }

    /// Initialize the test optimization library with a mock tracer and a working directory
    #[allow(dead_code)]
    pub fn init_mock_with_working_dir(working_dir: &str) -> bool {
        Self::init_with_values(
            LANGUAGE_NAME,
            RUNTIME_NAME,
            Self::runtime_version(),
            Some(working_dir),
            true,
        )
    }

    /// Initialize the test optimization library with specific values
    #[allow(dead_code)]
    pub fn init_with_values(
        language_name: impl AsRef<str>,
        runtime_name: impl AsRef<str>,
        runtime_version: impl AsRef<str>,
        working_directory: Option<impl AsRef<str>>,
        use_mock_tracer: bool,
    ) -> bool {
        #[cfg(target_os = "windows")]
        unsafe {
            // On Windows, call the platform-specific initialization
            // this is required on static libraries compiled by the go toolchain
            // just to start the go runtime
            _rt0_amd64_windows_lib()
        }

        // Create CStrings for the required parameters
        let language_name_cstring = CString::new(language_name.as_ref()).unwrap();
        let runtime_name_cstring = CString::new(runtime_name.as_ref()).unwrap();
        let runtime_version_cstring = CString::new(runtime_version.as_ref()).unwrap();
        // Create an optional CString for working_directory if provided
        let working_directory_cstring =
            working_directory.map(|wd| CString::new(wd.as_ref()).unwrap());

        // Build the initialization options struct, using as_ptr() so the memory is managed automatically
        let init_options = topt_InitOptions {
            language: language_name_cstring.as_ptr() as *mut c_char,
            runtime_name: runtime_name_cstring.as_ptr() as *mut c_char,
            runtime_version: runtime_version_cstring.as_ptr() as *mut c_char,
            working_directory: working_directory_cstring
                .as_ref()
                .map_or(null_mut(), |s| s.as_ptr() as *mut c_char),
            environment_variables: null_mut(),
            global_tags: null_mut(),
            use_mock_tracer: if use_mock_tracer { 1 } else { 0 },
            unused01: null_mut(),
            unused02: null_mut(),
            unused03: null_mut(),
            unused04: null_mut(),
            unused05: null_mut(),
        };

        // Initialize the library with the provided options
        unsafe { Bool_to_bool(topt_initialize(init_options)) }
    }

    /// Shutdown the test optimization library
    #[allow(dead_code)]
    pub fn shutdown() -> bool {
        unsafe { Bool_to_bool(topt_shutdown()) }
    }

    /// Get the current settings
    #[allow(dead_code)]
    pub fn get_settings() -> Settings {
        unsafe {
            let settings_response = topt_get_settings();
            Settings {
                code_coverage: Bool_to_bool(settings_response.code_coverage),
                early_flake_detection: EfDSettings {
                    enabled: Bool_to_bool(settings_response.early_flake_detection.enabled),
                    slow_test_retries: EfdSlowTestRetriesSettings {
                        ten_s: settings_response
                            .early_flake_detection
                            .slow_test_retries
                            .ten_s,
                        thirty_s: settings_response
                            .early_flake_detection
                            .slow_test_retries
                            .thirty_s,
                        five_m: settings_response
                            .early_flake_detection
                            .slow_test_retries
                            .five_m,
                        five_s: settings_response
                            .early_flake_detection
                            .slow_test_retries
                            .five_s,
                    },
                    faulty_session_threshold: settings_response
                        .early_flake_detection
                        .faulty_session_threshold,
                },
                flaky_test_retries_enabled: Bool_to_bool(
                    settings_response.flaky_test_retries_enabled,
                ),
                itr_enabled: Bool_to_bool(settings_response.itr_enabled),
                require_git: Bool_to_bool(settings_response.require_git),
                tests_skipping: Bool_to_bool(settings_response.tests_skipping),
                known_tests_enabled: Bool_to_bool(settings_response.known_tests_enabled),
                test_management: TestManagementSettings {
                    enabled: Bool_to_bool(settings_response.test_management.enabled),
                    attempt_to_fix_retries: settings_response
                        .test_management
                        .attempt_to_fix_retries,
                },
            }
        }
    }

    /// Get the flaky test retries settings
    #[allow(dead_code)]
    pub fn get_flaky_test_retries_settings() -> FlakyTestRetriesSettings {
        unsafe {
            let response = topt_get_flaky_test_retries_settings();
            FlakyTestRetriesSettings {
                retry_count: response.retry_count,
                total_retry_count: response.total_retry_count,
            }
        }
    }

    /// Get the known tests
    #[allow(dead_code)]
    pub fn get_known_tests() -> HashMap<String, HashMap<String, Vec<String>>> {
        unsafe {
            let mut modules_map: HashMap<String, HashMap<String, Vec<String>>> = HashMap::new();
            let known_tests = topt_get_known_tests();
            for i in 0..known_tests.len {
                let element = &*known_tests.data.add(i);

                let module_name_c = CStr::from_ptr(element.module_name);
                let suite_name_c = CStr::from_ptr(element.suite_name);
                let test_name_c = CStr::from_ptr(element.test_name);

                let module_name_string = module_name_c.to_string_lossy().into_owned();
                let suite_name_string = suite_name_c.to_string_lossy().into_owned();
                let test_name = test_name_c.to_string_lossy().into_owned();

                let suites_map = modules_map
                    .entry(module_name_string)
                    .or_insert_with(HashMap::new);
                let tests_vec = suites_map.entry(suite_name_string).or_insert_with(Vec::new);
                tests_vec.push(test_name);
            }
            topt_free_known_tests(known_tests);
            modules_map
        }
    }

    /// Get the skippable tests
    #[allow(dead_code)]
    pub fn get_skippable_tests() -> HashMap<String, HashMap<String, Vec<SkippableTest>>> {
        unsafe {
            let mut suites_map: HashMap<String, HashMap<String, Vec<SkippableTest>>> =
                HashMap::new();
            let skippable_tests = topt_get_skippable_tests();
            for i in 0..skippable_tests.len {
                let element = &*skippable_tests.data.add(i);

                let suite_name_c = CStr::from_ptr(element.suite_name);
                let test_name_c = CStr::from_ptr(element.test_name);
                let parameters_c = CStr::from_ptr(element.parameters);
                let custom_configurations_json_c =
                    CStr::from_ptr(element.custom_configurations_json);

                let suite_name_string = suite_name_c.to_string_lossy().into_owned();
                let test_name_string = test_name_c.to_string_lossy().into_owned();
                let parameters_string = parameters_c.to_string_lossy().into_owned();
                let custom_configurations_json_string =
                    custom_configurations_json_c.to_string_lossy().into_owned();

                let suites_map_entry = suites_map
                    .entry(suite_name_string.clone())
                    .or_insert_with(HashMap::new);
                let tests_vec = suites_map_entry
                    .entry(test_name_string.clone())
                    .or_insert_with(Vec::new);

                tests_vec.push(SkippableTest {
                    suite_name: suite_name_string,
                    test_name: test_name_string,
                    parameters: parameters_string,
                    custom_configurations_json: custom_configurations_json_string,
                });
            }
            topt_free_skippable_tests(skippable_tests);
            suites_map
        }
    }

    /// Get the test management tests
    #[allow(dead_code)]
    pub fn get_test_management_tests(
    ) -> HashMap<String, HashMap<String, HashMap<String, TestManagementTest>>> {
        unsafe {
            let mut modules_map: HashMap<
                String,
                HashMap<String, HashMap<String, TestManagementTest>>,
            > = HashMap::new();
            let test_management_tests = topt_get_test_management_tests();
            for i in 0..test_management_tests.len {
                let element = &*test_management_tests.data.add(i);

                let module_name_c = CStr::from_ptr(element.module_name);
                let suite_name_c = CStr::from_ptr(element.suite_name);
                let test_name_c = CStr::from_ptr(element.test_name);

                let module_name_string = module_name_c.to_string_lossy().into_owned();
                let suite_name_string = suite_name_c.to_string_lossy().into_owned();
                let test_name_string = test_name_c.to_string_lossy().into_owned();

                let modules_map_entry = modules_map
                    .entry(module_name_string.clone())
                    .or_insert_with(HashMap::new);
                let suites_map_entry = modules_map_entry
                    .entry(suite_name_string.clone())
                    .or_insert_with(HashMap::new);
                _ = suites_map_entry.entry(test_name_string.clone()).or_insert(
                    TestManagementTest {
                        module_name: module_name_string,
                        suite_name: suite_name_string,
                        test_name: test_name_string,
                        quarantined: Bool_to_bool(element.quarantined),
                        disabled: Bool_to_bool(element.disabled),
                        attempt_to_fix: Bool_to_bool(element.attempt_to_fix),
                    },
                );
            }
            topt_free_test_management_tests(test_management_tests);
            modules_map
        }
    }
}

File: sdks/python/test-optimization-sdk/src/test_optimization_sdk/test_optimization.py
"""
Test optimization module for managing the test optimization library.

This module provides functionality for initializing and shutdown the library.
Also access to the backend features.
"""

import platform
import sys
from dataclasses import dataclass
from typing import Dict, List, Optional

from .lib import (
    ffi,
    topt_initialize,
    topt_shutdown,
    topt_get_settings,
    topt_get_flaky_test_retries_settings,
    topt_get_known_tests,
    topt_get_skippable_tests,
    topt_get_test_management_tests,
    topt_free_known_tests,
    topt_free_skippable_tests,
    topt_free_test_management_tests,
    topt_test_set_source,
)


@dataclass
class Settings:
    """Represents the settings for a test session."""

    code_coverage: bool  # Whether code coverage is enabled
    early_flake_detection: "EfDSettings"  # Settings for early flake detection
    flaky_test_retries_enabled: bool  # Whether flaky test retries are enabled
    itr_enabled: bool  # Whether intelligent test runner is enabled
    require_git: bool  # Whether git integration is required
    tests_skipping: bool  # Whether test skipping is enabled
    known_tests_enabled: bool  # Whether known tests tracking is enabled
    test_management: "TestManagementSettings"  # Settings for test management


@dataclass
class EfDSettings:
    """Settings for early flake detection."""

    enabled: bool  # Whether early flake detection is enabled
    slow_test_retries: "EfdSlowTestRetriesSettings"  # Settings for slow test retries
    faulty_session_threshold: int  # Threshold for faulty session detection


@dataclass
class EfdSlowTestRetriesSettings:
    """Settings for slow test retries in early flake detection."""

    five_m: int  # Number of retries for 5-minute tests
    thirty_s: int  # Number of retries for 30-second tests
    ten_s: int  # Number of retries for 10-second tests
    five_s: int  # Number of retries for 5-second tests


@dataclass
class FlakyTestRetriesSettings:
    """Settings for flaky test retries."""

    retry_count: int  # Number of retries for flaky tests
    total_retry_count: int  # Total number of retries allowed


@dataclass
class TestManagementSettings:
    """Settings for test management."""

    enabled: bool  # Whether test management is enabled
    attempt_to_fix_retries: int  # Number of retries for attempt-to-fix operations
    __test__ = False  # Tell pytest to ignore this class


@dataclass
class SkippableTest:
    """Represents a skippable test."""

    suite_name: str  # Name of the test suite
    test_name: str  # Name of the test
    parameters: str  # Test parameters
    custom_configurations_json: str  # Custom configurations in JSON format


@dataclass
class TestManagementTest:
    """Represents a test managed by the test management system."""

    module_name: str  # Name of the module
    suite_name: str  # Name of the test suite
    test_name: str  # Name of the test
    quarantined: bool  # Whether the test is quarantined
    disabled: bool  # Whether the test is disabled
    attempt_to_fix: bool  # Whether the test is attempt-to-fix
    __test__ = False  # Tell pytest to ignore this class


class TestOptimization:
    """Represents a test session."""

    # Language name for the test session
    LANGUAGE_NAME = "python"
    # Runtime name for the test session
    RUNTIME_NAME = "python"

    @staticmethod
    def runtime_version() -> str:
        """Get the runtime version."""
        return f"{sys.version_info.major}.{sys.version_info.minor}.{sys.version_info.micro}"

    @classmethod
    def init(cls) -> bool:
        """Initialize the test optimization library."""
        return cls.init_with_values(
            cls.LANGUAGE_NAME,
            cls.RUNTIME_NAME,
            cls.runtime_version(),
            None,
            False,
        )

    @classmethod
    def init_with_working_dir(cls, working_dir: str) -> bool:
        """Initialize the test optimization library with a working directory."""
        return cls.init_with_values(
            cls.LANGUAGE_NAME,
            cls.RUNTIME_NAME,
            cls.runtime_version(),
            working_dir,
            False,
        )

    @classmethod
    def init_mock(cls) -> bool:
        """Initialize the test optimization library with a mock tracer."""
        return cls.init_with_values(
            cls.LANGUAGE_NAME,
            cls.RUNTIME_NAME,
            cls.runtime_version(),
            None,
            True,
        )

    @classmethod
    def init_mock_with_working_dir(cls, working_dir: str) -> bool:
        """Initialize the test optimization library with a mock tracer and a working directory."""
        return cls.init_with_values(
            cls.LANGUAGE_NAME,
            cls.RUNTIME_NAME,
            cls.runtime_version(),
            working_dir,
            True,
        )

    @classmethod
    def init_with_values(
        cls,
        language_name: str,
        runtime_name: str,
        runtime_version: str,
        working_directory: Optional[str],
        use_mock_tracer: bool,
    ) -> bool:
        """Initialize the test optimization library with specific values."""
        # Create C strings for the required parameters
        language_name_c = ffi.new("char[]", language_name.encode())
        runtime_name_c = ffi.new("char[]", runtime_name.encode())
        runtime_version_c = ffi.new("char[]", runtime_version.encode())
        working_directory_c = ffi.new("char[]", working_directory.encode()) if working_directory else ffi.NULL

        # Build the initialization options struct
        init_options = ffi.new("topt_InitOptions*", {
            "language": language_name_c,
            "runtime_name": runtime_name_c,
            "runtime_version": runtime_version_c,
            "working_directory": working_directory_c,
            "environment_variables": ffi.NULL,
            "global_tags": ffi.NULL,
            "use_mock_tracer": 1 if use_mock_tracer else 0,
            "unused01": ffi.NULL,
            "unused02": ffi.NULL,
            "unused03": ffi.NULL,
            "unused04": ffi.NULL,
            "unused05": ffi.NULL,
        })

        # Initialize the library with the provided options
        return bool(topt_initialize(init_options))

    @staticmethod
    def shutdown() -> bool:
        """Shutdown the test optimization library."""
        return bool(topt_shutdown())

    @staticmethod
    def get_settings() -> Settings:
        """Get the current settings."""
        settings = topt_get_settings()
        return Settings(
            code_coverage=bool(settings.code_coverage),
            early_flake_detection=EfDSettings(
                enabled=bool(settings.early_flake_detection.enabled),
                slow_test_retries=EfdSlowTestRetriesSettings(
                    five_m=settings.early_flake_detection.slow_test_retries.five_m,
                    thirty_s=settings.early_flake_detection.slow_test_retries.thirty_s,
                    ten_s=settings.early_flake_detection.slow_test_retries.ten_s,
                    five_s=settings.early_flake_detection.slow_test_retries.five_s,
                ),
                faulty_session_threshold=settings.early_flake_detection.faulty_session_threshold,
            ),
            flaky_test_retries_enabled=bool(settings.flaky_test_retries_enabled),
            itr_enabled=bool(settings.itr_enabled),
            require_git=bool(settings.require_git),
            tests_skipping=bool(settings.tests_skipping),
            known_tests_enabled=bool(settings.known_tests_enabled),
            test_management=TestManagementSettings(
                enabled=bool(settings.test_management.enabled),
                attempt_to_fix_retries=settings.test_management.attempt_to_fix_retries,
            ),
        )

    @staticmethod
    def get_flaky_test_retries_settings() -> FlakyTestRetriesSettings:
        """Get the current flaky test retries settings."""
        settings = topt_get_flaky_test_retries_settings()
        return FlakyTestRetriesSettings(
            retry_count=settings.retry_count,
            total_retry_count=settings.total_retry_count,
        )

    @staticmethod
    def get_known_tests() -> Dict[str, Dict[str, List[str]]]:
        """Get the list of known tests."""
        # Get the array from the native side
        known_array = topt_get_known_tests()
        try:
            # Convert the C array into a Dict[str, Dict[str, List[str]]]
            result = {}
            if known_array.data:
                for i in range(known_array.len):
                    test = known_array.data[i]
                    module_name = test.module_name.decode() if test.module_name else ""
                    suite_name = test.suite_name.decode() if test.suite_name else ""
                    test_name = test.test_name.decode() if test.test_name else ""

                    if module_name not in result:
                        result[module_name] = {}
                    if suite_name not in result[module_name]:
                        result[module_name][suite_name] = []
                    result[module_name][suite_name].append(test_name)

            return result
        finally:
            # Free the native array
            topt_free_known_tests(known_array)

    @staticmethod
    def get_skippable_tests() -> Dict[str, Dict[str, List[SkippableTest]]]:
        """Get the list of skippable tests."""
        # Get the array from the native side
        skippable_array = topt_get_skippable_tests()
        try:
            # Convert the C array into a Dict[str, Dict[str, List[SkippableTest]]]
            result = {}
            if skippable_array.data:
                for i in range(skippable_array.len):
                    test = skippable_array.data[i]
                    suite_name = test.suite_name.decode() if test.suite_name else ""
                    test_name = test.test_name.decode() if test.test_name else ""

                    if suite_name not in result:
                        result[suite_name] = {}
                    if test_name not in result[suite_name]:
                        result[suite_name][test_name] = []

                    result[suite_name][test_name].append(
                        SkippableTest(
                            suite_name=suite_name,
                            test_name=test_name,
                            parameters=test.parameters.decode() if test.parameters else "",
                            custom_configurations_json=test.custom_configurations_json.decode()
                            if test.custom_configurations_json
                            else "",
                        )
                    )

            return result
        finally:
            # Free the native array
            topt_free_skippable_tests(skippable_array)

    @staticmethod
    def get_test_management_tests() -> Dict[str, Dict[str, Dict[str, TestManagementTest]]]:
        """Get the list of test management tests."""
        # Get the array from the native side
        test_array = topt_get_test_management_tests()
        try:
            # Convert the C array into a Dict[str, Dict[str, Dict[str, TestManagementTest]]]
            result = {}
            if test_array.data:
                for i in range(test_array.len):
                    test = test_array.data[i]
                    module_name = test.module_name.decode() if test.module_name else ""
                    suite_name = test.suite_name.decode() if test.suite_name else ""
                    test_name = test.test_name.decode() if test.test_name else ""

                    if module_name not in result:
                        result[module_name] = {}
                    if suite_name not in result[module_name]:
                        result[module_name][suite_name] = {}
                    if test_name not in result[module_name][suite_name]:
                        result[module_name][suite_name][test_name] = TestManagementTest(
                            module_name=module_name,
                            suite_name=suite_name,
                            test_name=test_name,
                            quarantined=bool(test.quarantined),
                            disabled=bool(test.disabled),
                            attempt_to_fix=bool(test.attempt_to_fix),
                        )

            return result
        finally:
            # Free the native array
            topt_free_test_management_tests(test_array)
